AI #1630: Document recent decisions by the TF-RM on naming and numbering

mike-finopsorg opened the pull request:
Signed-off-by: Mike Fuller <mike@finops.org>

--

# PR Summary
(added by @shawnalpay, AI-assisted)

## Overview
Update the **FOCUS writing guidelines** to document decisions made by the Requirements Model Task Force regarding **naming conventions**, **numbering**, and **entity structure**, specifically focusing on how to handle **multi-dataset** scenarios.

## Key Changes
The primary changes are to `guidelines/writing-requirements-model-guidelines.md`:

1.  **Entity Structure & Naming**:
    * Documented decisions regarding **multi-dataset entity structure**.
    * Updated **naming and numbering** conventions for the spec.
2.  **Resource Model (RM) Keys**:
    * Added details about a new **"Order" key** in the Resource Model.
    * Added **"Dataset" keys** to the guidelines.
3.  **Diagrams**:
    * Updated the **entity diagram** to reflect structural changes.

mike-finopsorg committed:
Document recent decisions by the TF-RM on naming and numbering
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
initial stage1 file for the Requirements Model updates to 1.3
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -137,8 +138,49 @@ This ensures traceability, uniqueness, and clarity.
+| Dataset | Abbreviation | Description |
+|---------|--------------|-------------|
+| Cost and Usage | CAU | Primary billing data with usage and cost information |
+| Contract Commitment | CC | Commitment discount and reservation data |
```
CCT

mike-finopsorg committed:
Requirements Model updates
Update all existing Column and Dataset Entities to have the CAU- prefix
Migrated PricingCurrencyListUnitPrice to 1.3
Updated unit test to support new CAU- prefix
Updated schema to support prefix
Updated schama to allow new Order key
Upated summary.py to support ordering
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Fixed ordering of dependencies on pricingcurrencylistunitprice
added unit test to enforce dependency order
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added details in the guidelines about the new "Order" key in the RM
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added entity options for DatasetId and DatasetName
Updated PricingCurrencyListUnitPrice to use new Dataset keys
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added dataset keys to RM guidelines
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Updated sumary.py to be output_normative_text_from_model.py  and added more features for use
Added docs in README.md on use of output_normative_text_from_model.py
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added initial list of stage1 files for sample columns as per #1629
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
1.3 Requirements Model updates for AvailabilityZone
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added RequirementsModel changes to CommitmentDiscountQuantity for FOCUS 1.3
Updated output script to support nesting rules to help with comparisons
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Updated requirements model for ColumnHandling to match 1.3
Signed-off-by: Mike Fuller <mike@finops.org>

jpradocueva committed:
updates

jpradocueva committed:
add three new columns for contract commitment

jpradocueva committed:
add contract commitment description

jpradocueva committed:
add contract commitment id

jpradocueva committed:
updates

jpradocueva committed:
add contract commitment period end

jpradocueva committed:
add contract commitment period start

jpradocueva committed:
add contract commitment quantity

jpradocueva committed:
add contract commitment type

jpradocueva committed:
add contract commitment unit

jpradocueva committed:
add contractid

jpradocueva committed:
add contract period end

jpradocueva committed:
add contract period start

jpradocueva committed:
add dataset

jpradocueva committed:
updates

jpradocueva committed:
updates

jpradocueva committed:
add diff for several cost and usage

jpradocueva committed:
add billing 7 invoice columns

jpradocueva committed:
add new list of columns on cost and usage

jpradocueva committed:
add allocatedmethoddetails, allocatedmethodid, allocatedresourceid

jpradocueva committed:
updates to allocatedmethoddetails

jpradocueva committed:
update allocatedmethodid

jpradocueva committed:
updates

jpradocueva committed:
updates

jpradocueva committed:
updates to allocated resource name

jpradocueva committed:
updates

jpradocueva committed:
updates

jpradocueva committed:
updates

jpradocueva committed:
add dataset cost and usage

jpradocueva committed:
update folder structure

jpradocueva committed:
add datatet

jpradocueva committed:
add data generator

jpradocueva committed:
add data generator overview

jpradocueva committed:
add datase instance id

jpradocueva committed:
add dataset instance name

jpradocueva committed:
add dataset instance

jpradocueva committed:
updates

jpradocueva committed:
add dataset instance complete

jpradocueva committed:
updates

jpradocueva committed:
updates

jpradocueva committed:
add dataset instance id

jpradocueva committed:
add dataset instance last updated

jpradocueva committed:
add recency last updated

jpradocueva committed:
add recency overivew

jpradocueva committed:
add billedcost

jpradocueva committed:
add listcost

jpradocueva committed:
add effective cost

jpradocueva committed:
updates to v1.3 without cr

jpradocueva committed:
add contracted cost

jpradocueva committed:
add consumed quantity

jpradocueva committed:
add list unit price

jpradocueva committed:
add Contracted Unit Price

jpradocueva committed:
add pricing category

jpradocueva committed:
add pricing currency

jpradocueva committed:
add pricing currency contracted unit price

jpradocueva committed:
add pricing currency effective cost

jpradocueva committed:
add Pricing Quantity

jpradocueva committed:
add SKU Price Details

jpradocueva committed:
add chargecategory

jpradocueva committed:
add charge Class

jpradocueva committed:
add charged description

jpradocueva committed:
add chargefrequency

jpradocueva committed:
add charge period start

jpradocueva committed:
add chargeperiodend

jpradocueva committed:
add service category

jpradocueva committed:
add service subcategory

jpradocueva committed:
add servicename

jpradocueva committed:
add resourceid

jpradocueva committed:
add resource name

jpradocueva committed:
add resource type

jpradocueva committed:
add sku id

jpradocueva committed:
add sku meter;

jpradocueva committed:
add tags

jpradocueva committed:
skui

jpradocueva committed:
add metadat overview

jpradocueva committed:
add time sector complete

jpradocueva committed:
add time sector end

jpradocueva committed:
add time sector last updated

jpradocueva committed:
add timesectorstart

jpradocueva committed:
add time sector overview

jpradocueva committed:
add ColumnDefinition

jpradocueva committed:
add datasetinstanceId

jpradocueva committed:
add schema

jpradocueva committed:
updates

jpradocueva committed:
updates

jpradocueva committed:
add discount handling

jpradocueva committed:
add string handling attribute

jpradocueva committed:
add unit format attribute

jpradocueva committed:
add discount handling

jpradocueva committed:
add invoice handling attribute

jpradocueva committed:
add json object format

jpradocueva committed:
add contract applied

mike-finopsorg committed:
Bump model version in requirements model
#1755 updates for CostAndUsage.Tags to suit FOCUS 1.3
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
#1754 CostAndUsage.SubAccountType to suit FOCUS 1.3
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.SkuId to suit FOCUS 1.3 #1748
CostAndUsage.SkuMeter to suit FOCUS 1.3 #1749
CostAndUsage.SkuPriceDetails to suit FOCUS 1.3 #1750
CostAndUsage.SkuPriceId to suit FOCUS 1.3 #1751
CostAndUsage.SubAccountId to suit FOCUS 1.3 #1752
Further enhancements to helper script
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.ServiceSubcategory to suit FOCUS 1.3 #1747
CostAndUsage.ServiceName to suit FOCUS 1.3 #1745
CostAndUsage.ServiceCategory to suit FOCUS 1.3 #1744
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.ResourceId to suit FOCUS 1.3 #1741
CostAndUsage.ResourceName to suit FOCUS 1.3 #1742
CostAndUsage.ResourceType to suit FOCUS 1.3 #1743
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.RegionName to suit FOCUS 1.3 #1740
CostAndUsage.RegionId to suit FOCUS 1.3 #1739
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.BilledCost to suit FOCUS 1.3 #1697
CostAndUsage.BillingAccountId to suit FOCUS 1.3 #1698
CostAndUsage.BillingAccountName to suit FOCUS 1.3 #1699
CostAndUsage.BillingAccountType to suit FOCUS 1.3 #1700
CostAndUsage.BillingCurrency to suit FOCUS 1.3 #1701
CostAndUsage.BillingPeriodEnd to suit FOCUS 1.3 #1702
CostAndUsage.BillingPeriodStart to suit FOCUS 1.3 #1703
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.CapacityReservationId to suit FOCUS 1.3 #1704
CostAndUsage.CapacityReservationStatus to suit FOCUS 1.3 #1705
CostAndUsage.ChargeCategory to suit FOCUS 1.3 #1706
CostAndUsage.ChargeClass to suit FOCUS 1.3 #1707
CostAndUsage.ChargeDescription to suit FOCUS 1.3 #1708
CostAndUsage.ChargeFrequency to suit FOCUS 1.3 #1709
CostAndUsage.ChargePeriodEnd to suit FOCUS 1.3 #1710
CostAndUsage.ChargePeriodStart to suit FOCUS 1.3 #1711

Update ordering test and fix out of order dependencies
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.PublisherName to suit FOCUS 1.3 #1738
CostAndUsage.ProviderName to suit FOCUS 1.3 #1737
CostAndUsage.PricingUnit to suit FOCUS 1.3 #1736
CostAndUsage.PricingQuantity to suit FOCUS 1.3 #1735
CostAndUsage.PricingCurrencyEffectiveCost to suit FOCUS 1.3 #1733
CostAndUsage.PricingCurrencyContractedUnitPrice to suit FOCUS 1.3 #1732
CostAndUsage.PricingCurrency to suit FOCUS 1.3 #1731
CostAndUsage.PricingCategory to suit FOCUS 1.3 #1730
CostAndUsage.ListUnitPrice to suit FOCUS 1.3 #1729
CostAndUsage.ListCost to suit FOCUS 1.3 #1728
CostAndUsage.InvoiceId to suit FOCUS 1.3 #1726
CostAndUsage.EffectiveCost to suit FOCUS 1.3 #1724
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.CommitmentDiscountCategory to suit FOCUS 1.3 #1712
CostAndUsage.CommitmentDiscountId to suit FOCUS 1.3 #1713
CostAndUsage.CommitmentDiscountName to suit FOCUS 1.3 #1714
CostAndUsage.CommitmentDiscountStatus to suit FOCUS 1.3 #1716
CostAndUsage.CommitmentDiscountType to suit FOCUS 1.3 #1717
CostAndUsage.CommitmentDiscountUnit to suit FOCUS 1.3 #1718
CostAndUsage.ConsumedQuantity to suit FOCUS 1.3 #1719
CostAndUsage.ConsumedUnit to suit FOCUS 1.3 #1720
CostAndUsage.ContractedCost to suit FOCUS 1.3 #1722
CostAndUsage.ContractedUnitPrice to suit FOCUS 1.3 #1723

Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.InvoiceIssuerName to suit FOCUS 1.3 #1727
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
CostAndUsage.SkuPriceId to suit FOCUS 1.3
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
attribute.CurrencyFormat to suit FOCUS 1.3 #1771
attribute.DatetimeFormat to suit FOCUS 1.3 #1773
attribute.DiscountHandling to suit FOCUS 1.3 #1774
attribute.KeyValueFormat to suit FOCUS 1.3 #1777
attribute.NullHandling to suit FOCUS 1.3 #1778
attribute.NumericFormat to suit FOCUS 1.3 #1779
attribute.StringHandling to suit FOCUS 1.3 #1798
attribute.UnitFormat to suit FOCUS 1.3 #1799

Signed-off-by: Mike Fuller <mike@finops.org>

jpradocueva committed:
AI #1652: Remove composite children from root requirement (#1795)

As we discussed during the Requirements Meeting on December 9. The root
requirement should include only the composite requirements, not their
children. This PR is removing the composite children from the root
requirement.

### Questions for @mike-finopsorg 
- [ ] In `ContractApplied` column, `CAU-ContractApplied-C-013-C` and
`CAU-ContractApplied-C-017-M` are children of
`CAU-ContractApplied-C-004-C`, but at the same time, they are composites
that have their own children. They represent the second nested level.
Should they be represented in the root requirement,
`CAU-ContractApplied-C-000-C`? At the moment they are. These also
applied to the following criteria, e.g., `-C-026-C`, `-C-029-C`,
`-C-032-C`, `-C-036-C`, `-C-039-C`, `-C-045-C`, `-C-049-C`, `-C-054-C`,
`-C-058-C`, `-C-063-C`, and `-C-068-C`. In short, these requirements
appear twice: once at the root and once within their parents. I believe
that the second level should be removed, but I will wait for your
feedback.

mike-finopsorg committed:
Updated after todays TF-RM deprecated rules remain linked.
Added new status of "Removed" allowing the model to track removed rules.
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Updated schema to allow EntityType Object
Updated schema to require ModelVersionRemoved when Status: Removed and not allow ModelVersionRemoved in other cases

Signed-off-by: Mike Fuller <mike@finops.org>

shawnalpay commented:
@mike-finopsorg to incorporate feedback from @jpradocueva.

I won't bring this to Members for review until I see an approval from @jpradocueva.

mike-finopsorg committed:
Added new schema elements
DatasetType
EntityId
EntityName
Updated to support new EntityType Object
Added AllocatedMethodDetails and AllocatedMethodDetailsObject
Updated unittests
New ApplicabilityCriteria and

Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Force the requirement for Order key
Fix issues with output normative text script
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added remaining Allocated* columns to 1.3 Model
Added extra test to detect duplicate Order
Tidy up Order values
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added allocatedTags
Added contractapplied
New checkfunctions
Fixed order numbering in allocatedmethodetailsobject
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added HostProviderName
Added ServiceProviderName
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Update BilledCost now ServiceProviderName is in the model
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Updated the guidelines for requirements model to suit recent updates
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg commented:
Applied updates as per the recent updates. Awaiting @jpradocueva to review and add his suggestions.

mike-finopsorg committed:
Restructure folders for multiple-datasets
Added contract commitments dataset
Added ContractCommitmentPeriodEnd column
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Restructure folders for multiple-datasets
Added contract commitments dataset
Added ContractCommitmentPeriodEnd column
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Fix spacing in json files
Added tests for RM
Update script output_normative_text_from_model to support multi dataset
Added columns for contractcommitment dataset
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Added attributes invoicehandling and jsonobjectformat to requirements model
Signed-off-by: Mike Fuller <mike@finops.org>

jpradocueva committed:
enhancements to align the model with the currenct practices

jpradocueva reviewed (changes requested):
@mike-finopsorg, I have created PR #1905 with enhancements to the model based on the JSON properties and extraction process. Please let me know if you have any comments.

mike-finopsorg committed:
Adjusted status of rule in columnhandling to use new Removed status
Added attribute datageneratorcalculatedsplitcostalocation
updated stringhandling for 1.3 rules
Updates to tests and corrections to helper script to support removed status
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Merge branch 'working_draft' into 1577-ai-add-all-13-spec-changes-to-the-requirements-model

mike-finopsorg committed:
Merge branch 'working_draft' into 1577-ai-add-all-13-spec-changes-to-the-requirements-model

mike-finopsorg committed:
Requirements Model corrections:
Unify the way we represent the entity in the DateTimeFormat attribute
Remove incorrect rule in ContractCommitmentCategory
Correct language in ContractAppliedObject to match spec
Correct presence rule wording for PublisherName and ProviderName
Signed-off-by: Mike Fuller <mike@finops.org>

ijurica commented on guidelines/writing-requirements-model-guidelines.md:
````diff
@@ -56,8 +58,6 @@ Identify the target for the rule: **Dataset**, **Column**, **Attribute** propert
 
-**FOCUS Architectural components**
-
 ```mermaid
````
Suggestion: Add a relationship between Dataset and Attribute.

shawnalpay replied:
@jpradocueva Please address this with a commit.

thecloudman committed:
Fix typo in allocatedmethodid.json key

Change case of Id to ID to match spec

mike-finopsorg committed:
Correct name of CAU-AllocatedMethodID-C-000-C to CAU-AllocatedMethodId-C-000-C
Update dependencies to avoid circular references
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Correct names of files as per feedback
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Merge pull request #1905 from FinOps-Open-Cost-and-Usage-Spec/1630-dataset-entity-structure-naming-enhancements

AI #1630: Enhancements to align the model with the current practices

shawnalpay commented:
@jpradocueva 
Now that we have merged PR #1905 into this PR, I have re-requested your review.

@mike-finopsorg 
Could you please:
* Resolve the merge conflict?
* Add a summary to the parent post?
* Let me know if we can advance this to Member review (hopefully once Joaquin approves)?  And if not (presumably because there is still more to add) -- shall we perhaps put it back to draft status?

mike-finopsorg committed:
Merge branch 'working_draft' into 1577-ai-add-all-13-spec-changes-to-the-requirements-model

mike-finopsorg committed:
Corrections to allocatedmethoddetailsobject missing connection to rules
Fix contractappliedobject condition incorrect checkfunction
Signed-off-by: Mike Fuller <mike@finops.org>

jpradocueva committed:
Add the missing word "service" to CAU-ResourceId-C-007-M

Add the missing word "service" to CAU-ResourceId-C-007-M

mike-finopsorg committed:
Merge branch 'working_draft' into 1630-ai-document-the-decisions-about-multi-dataset-entity-structure-and-naming

jpradocueva committed:
update entity diagram

jpradocueva reviewed (approved):
The old entity diagram (Mermaid) was removed by PR #1905. The new entity diagram is a placeholder that will evolve with time, e.g., the addition of Metadata.

shawnalpay commented:
A TF approval and no open threads.  Graduating to `PR Member Review`!

mike-finopsorg committed:
Refinements based on feedback given
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Merge branch 'working_draft' into 1577-ai-add-all-13-spec-changes-to-the-requirements-model

mike-finopsorg committed:
Remove extra word not found in spec.
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Correct ordering of items in BillingAccountId
Signed-off-by: Mike Fuller <mike@finops.org>

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -24,90 +24,61 @@ An Action Item (AI) ticket should be opened to track the progress of implementin
 | `Attribute`        | Shared formatting/logic constraint                  | Formatting consistency across columns     | All `String` columns MUST conform to `StringHandling` requirements                                              |
 | `Metadata`         | Schema-level dataset descriptors                    | Schema versioning, declaration            | Metadata MUST declare `focus_version` as a valid semantic version string (e.g., "1.2.0")                        |
-| `Service Provider` | Entity that made the resource available for purchase | Conditional logic in requirements         | Column `CapacityReservationId` MUST be present when the service provider supports capacity reservation features |
+| `Object`           | TBD                                                 | TBD                                       | TBD                                                                                                             |
```
Please replace `TBD` with content.

TechinMama replied:
It also looks like it might be a new or changed concept replacing something else. Should this be notated earlier on in the file, within the repository for folks expecting the former?

shawnalpay replied:
Looks good now.  Resolving!

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -24,90 +24,61 @@ An Action Item (AI) ticket should be opened to track the progress of implementin
-|-------------------|----------|-----------|-----------------------|-----------|-------------|---------|-------------|------|-------|-----|----|
-| | | | | | | | | | | | |
-
 - `ModelRuleId` - Is the formal Id given to this entry in the model rules (See: [CR Expression Format](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/1121-ai-align-on-approach-for-scrs/specification/requirements_model/README.md#-cr-expression-format))
```
Please scrub any presence of "CR" to "RM".

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -24,90 +24,61 @@ An Action Item (AI) ticket should be opened to track the progress of implementin
-**FOCUS Entity Reference Table**
+##### FOCUS Entity Reference Table
 
 | Entity             | Description                                         | Applies To                                | Example CR Function                                                                                             |
```
Please scrub any use of "CR" to "RM".

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -137,15 +106,54 @@ This ensures traceability, uniqueness, and clarity.
-**Reasoning Rules**
-
 - Use `Presence` for rules requiring the column’s inclusion in the dataset.
 - Use `DataType` to enforce primitive types like `Decimal`, `String`, `Boolean`.
```
Should this be `Type` instead of `DataType`...?

mike-finopsorg replied:
```suggestion
- Use `Type` to enforce primitive types like `Decimal`, `String`, `Boolean`.
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -137,15 +106,54 @@ This ensures traceability, uniqueness, and clarity.
-
 - Use `Presence` for rules requiring the column’s inclusion in the dataset.
 - Use `DataType` to enforce primitive types like `Decimal`, `String`, `Boolean`.
 - Use `Format` for pattern-based constraints (e.g., `DateTimeFormat`, `UUID`, `NumericFormat`).
```
In the next line, you mention `NullabilityRules`, but it's never used in the examples.... should we strike?

mike-finopsorg replied:
Updated to correct Nullability

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
This is not in scope for the PR as proposed, but I feel that you should go back to the top of the document and take a stab at rewriting it.  Make explicit use of the term `Requirements Model`.  And when you say "rules", make it clear that you're describing it in the scope of the Requirements Model.  I could see someone reading this and thinking you're talking about normative requirements in the spec.

# Guidelines for writing model rules

As the FOCUS working group moves to introduce a formal rule definition structure the requirement for FOCUS members to understand how to read and write rules will increase. This guide is here to assist those who are starting their journey in writing rules for FOCUS.

The `specification/requirements_model` folder contains modular model components and a Python-based build process that assembles them into a validated `model-<version>.json` file using a corresponding JSON Schema (`model_schema.json`).

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -265,8 +261,6 @@ A rule states: “The following rules MUST be enforced for `CommitmentDiscountQu
-
 - Use `static` if the rule can be enforced by examining the dataset alone:  
   Example: value types, nullability, formatting, or schema presence.
 - Use `dynamic` if validation depends on:
```
Fix casing.  Should be Proper Case.

```suggestion
- Use `Static` if the rule can be enforced by examining the dataset alone:  
  Example: value types, nullability, formatting, or schema presence.
- Use `Dynamic` if validation depends on:
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -196,8 +200,6 @@ A rule states: “Rows SHOULD include `SkuId` when `ChargeCategory = Purchase`.
 
-**Reasoning Rules**
-
 - Use `"All_Rows"` when no structural gating is defined.
```
Is this right?  Seems like it should be:

```suggestion
- Use an empty list [] when no structural gating is defined.
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -24,90 +24,61 @@ An Action Item (AI) ticket should be opened to track the progress of implementin
+- `Order` - The order in which this rule should be processed or displayed
+- `DatasetType` - The dataset type this rule applies to (e.g. "CAU" for Cost and Usage, "CCT" for Contract Commitment)
+- `DatasetId` - The identifier of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "CostAndUsage" for Cost and Usage, "ContractCommitment" for Contract Commitment)
+- `DatasetName` - The human-readable name of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "Cost and Usage" for Cost and Usage,
```
```suggestion
- `DatasetName` - The human-readable name of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "Cost and Usage" for Cost and Usage)
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
````diff
@@ -573,15 +650,18 @@ Common rule for columns with a MUST be one of the allowed values requirement.
 ```json
-  "SampleColumn-C-007-M": {
+  "CAU-SampleColumn-C-007-M": {
     "Function": "Type",
````
Should this be Format...?

```suggestion
    "Function": "Format",
```

TechinMama commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -137,15 +106,54 @@ This ensures traceability, uniqueness, and clarity.
+| Dataset | Abbreviation | Description |
+|---------|--------------|-------------|
+| Cost and Usage | CAU | Primary billing data with usage and cost information |
+| Contract Commitment | CCT | Commitment discount and reservation data |
```
Here, CCT is used for Contract Commitment. This is the same for earlier in the file, "Line 40". Should this abbreviation align with the CRIDs in lines 91, 137, 141, and 142?

mike-finopsorg replied:
Corrected to CCT

TechinMama commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -358,22 +364,74 @@ If you need assistance reach out to Mike Fuller in the FOCUS slack
+
+**Best Practices:**
+
+- Use multiples of 10 (10, 20, 30...) to allow for future rule insertion
```
This looks like a new concept as well. In line 383, its mentioned to "use incremental values e.g." but sounds as if its possible to merely select incremental values. Here, in best practices, it mentions "use multiples of 10", which sounds more explicit. Similar wording should be used the same, or explicitly use multiples of 10 to avoid confusion. 

Havent' seen this at work, but would adding in a new rule "based on the example of Order: 5 and Order: 15, mean a future rule entry would follow "Order: 10"?

mike-finopsorg replied:
Cleaned up the example and stated that Order is no longer optional.

TechinMama commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -358,22 +364,74 @@ If you need assistance reach out to Mike Fuller in the FOCUS slack
     "Status": "Active",
     "ApplicabilityCriteria": [],
     "Type": "Static",
+    "Order": 0,
```
The introduction of ordering mentions that its optional. If not used, is "Order: 0" the correct entry or is this similar to 0 indexing?

jpradocueva committed:
AI 1577:  Remove requirements_model folder inside supporting_content (#2035)

The root-level directory `/supporting_content/requirements_model/`
contains .json artifacts generated during the initial manual extraction
of Markdown `(.md)` sources into a structured JSON format.

The finalized .json artifacts have since been consolidated under
`/specification/requirements_model/`. As a result, the contents of
`/supporting_content/requirements_model/` are redundant and no longer
required. All relevant requirement artifacts are now maintained within
`/specification/requirements_model/` folder.

mike-finopsorg committed:
Merge branch '1577-ai-add-all-13-spec-changes-to-the-requirements-model' into 1630-ai-document-the-decisions-about-multi-dataset-entity-structure-and-naming

mike-finopsorg committed:
Update guidelines/writing-requirements-model-guidelines.md

Co-authored-by: Shawn Alpay <shawn@finops.org>

mike-finopsorg committed:
Update guidelines/writing-requirements-model-guidelines.md

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -145,21 +107,60 @@ This ensures traceability, uniqueness, and clarity.
-- Use `DataType` to enforce primitive types like `Decimal`, `String`, `Boolean`.
+- Use `Type` to enforce primitive types like `Decimal`, `String`, `Boolean`.
 - Use `Format` for pattern-based constraints (e.g., `DateTimeFormat`, `UUID`, `NumericFormat`).
 - Use `NullabilityRules` to define when values must or must not be null.
```
```suggestion
- Use `Nullability` to define when values must or must not be null.
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -255,33 +246,29 @@ A rule states: “`BillingPeriodStart` MUST be of type `DateTime`.”
   - Catalog metadata  
   - Service Provider configuration or billing systems
-- For composite rules, set to `dynamic` if any child CRItem is dynamic.
+- For composite rules, set to `dynamic` if any child RM Item is dynamic.
```
```suggestion
- For composite rules, set to `Dynamic` if any child RM Item is dynamic.
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -302,13 +287,11 @@ Record the version of the FOCUS specification in which this rule was introduced.
-Indicate whether the rule is `active`, `deprecated`, or `reserved` for future use.
-
-**Reasoning Rules**
+Indicate whether the rule is `active`, `deprecated`, or `removed` for future use.
```
```suggestion
Indicate whether the rule is `Active`, `Deprecated`, or `Removed` for future use.
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -302,13 +287,11 @@ Record the version of the FOCUS specification in which this rule was introduced.
+Indicate whether the rule is `active`, `deprecated`, or `removed` for future use.
 
 - Default to `active` unless the normative text explicitly states otherwise.
 - Use `deprecated` if the rule is marked for removal or obsolescence.
```
```suggestion
- Default to `Active` unless the normative text explicitly states otherwise.
- Use `Deprecated` if the rule is marked for removal or obsolescence.
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -302,13 +287,11 @@ Record the version of the FOCUS specification in which this rule was introduced.
 - Default to `active` unless the normative text explicitly states otherwise.
 - Use `deprecated` if the rule is marked for removal or obsolescence.
-- Use `reserved` if the rule is included for future development or placeholder purposes.
+- Use `removed` if the rule is removed from the model.
```
```suggestion
- Use `Removed` if the rule is removed from the model.
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -366,14 +365,63 @@ If you need assistance reach out to Mike Fuller in the FOCUS slack
+- Use incremental values (e.g., 10, 20, 30) to allow for future insertions
+- Lower values indicate higher priority/earlier processing
+- Rules without an `Order` field are ignored for ordering purposes
+- The `Order` field is optional but recommended for rules that need explicit sequencing
```
```suggestion
- The `Order` field is required for rules to show explicit sequencing
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -366,14 +365,63 @@ If you need assistance reach out to Mike Fuller in the FOCUS slack
+    "ValidationCriteria": {
+      "Dependencies": [
+        "CAU-OtherRule-C-001-M",    // Order: 5
+        "CAU-AnotherRule-C-002-M"   // Order: 15
```
```suggestion
        "CAU-OtherRule-C-001-M",    // Order: 20
        "CAU-AnotherRule-C-002-M"   // Order: 30
```

mike-finopsorg committed:
Apply suggestions from code review

Co-authored-by: Shawn Alpay <shawn@finops.org>

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -145,21 +107,60 @@ This ensures traceability, uniqueness, and clarity.
+##### Examples
+
+- Cost and Usage dataset: `CAU-BillingAccountName-C-000-M`
+- Contract Commitment dataset: `CC-CommitmentDiscountId-C-001-M`
```
```suggestion
The `NNN` numbering is only unique within each dataset namespace. For example, `CAU-ListPrice-C-000-M` and `CCT-ListPrice-C-000-M` can both exist independently - the `000` is reused and only needs to be unique within that specific dataset abbreviation.

##### Examples

- Cost and Usage dataset: `CAU-BillingAccountName-C-000-M`
- Contract Commitment dataset: `CCT-CommitmentDiscountId-C-001-M`
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -24,118 +24,80 @@ An Action Item (AI) ticket should be opened to track the progress of implementin
-
-- Use the format: `ColumnID-EntityType-NNN-Level`
+- Use the format: `DatasetType-ColumnID-EntityType-NNN-Level`
+- `DatasetType`: Short identifier for the dataset (e.g., `CAU`, `CC`, `META`)
```
```suggestion
- `DatasetType`: Short identifier for the dataset (e.g., `CAU`, `CCT`)
```

mike-finopsorg committed:
Apply suggestions from code review

mike-finopsorg committed:
content refresh to align with current practices

mike-finopsorg committed:
Clearer that rule is a programic definition of a normative line.
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg commented:
I have cleaned this up a lot, feeling more confident on the content. @jpradocueva please review and let @shawnalpay know when we are ready to take this to members for approval.

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,82 @@ The model document for FOCUS contains the following major sections:
 #### 1. Target Entity – Determine the entity
 
-Identify the target for the rule: **Dataset**, **Column**, **Attribute** property, **Service Provider Name**, etc. This sets the scope of the model requirement.
+Identify the target for the rule: **Dataset**, **Column**, **Attribute** property, **Metadata**, **Object**etc. This sets the scope of the model requirement.
```
Small typo.  But further, could we make this a complete list and eliminate the etc?  It's these five, and only these five, yes?

```suggestion
Identify the target for the rule: **Dataset**, **Column**, **Attribute**, **Metadata**, or **Object**. This sets the scope of the model requirement.
```

mike-finopsorg replied:
Fixed

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,82 @@ The model document for FOCUS contains the following major sections:
+- `DatasetType` - The dataset type this rule applies to (e.g. "CAU" for Cost and Usage, "CCT" for Contract Commitment)
+- `DatasetId` - The identifier of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "CostAndUsage" for Cost and Usage, "ContractCommitment" for Contract Commitment)
+- `DatasetName` - The human-readable name of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "Cost and Usage" for Cost and Usage)
+- `ValidationCriteria` - The detailed criteria that defines how this rule is to be validated,
```
```suggestion
- `ValidationCriteria` - The detailed criteria that defines how this rule is to be validated
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -462,6 +674,7 @@ Common rule for columns with a NOT NULL requirement. Can also be used when there
     "DatasetId": "CostAndUsage",
     "DatasetName": "Cost and Usage",
     "Type": "Static",
+    "Order": 10,
```
The order on this line, and all subsequent lines in this example, should be incremented.  They're all currently 10.

mike-finopsorg replied:
I went back and forth on this. I'll make them increment.

mike-finopsorg replied:
```suggestion
    "Order": 20,
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,82 @@ The model document for FOCUS contains the following major sections:
+- `Notes` - Free form notes (short) included in the model rule document
+- `ModelVersionIntroduced` - Requirement Model Version this rule was added to the Model Rules
+- `Status` - Status of the rule (Valid values: Active, Deprecated, Removed)
+- `ModelVersionRemoved` - Requirement Model Version this rule was removed from the Model Rules
```
```suggestion
- `ModelVersionIntroduced` - Requirements Model Version this rule was added to the Model Rules
- `Status` - Status of the rule (Valid values: Active, Deprecated, Removed)
- `ModelVersionRemoved` - Requirements Model Version this rule was removed from the Model Rules
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -302,69 +330,251 @@ Record the version of the FOCUS specification in which this rule was introduced.
 - `model_rules/attributes/`: JSON files defining multiple `ModelRules` for a single attribute
 - `model_rules/columns/`: JSON files defining multiple `ModelRules` for a single column
 
+**Build output** (top-level):
```
Line 439 above (can't suggest an edit on those lines due to GitHub annoyingly not allowing suggestions on unchanged lines) I think needs to be:

> - model_rules/datasets/<dataset_id>/columns/: JSON files defining multiple `ModelRules` for a single column within a dataset
> - model_rules/datasets/<dataset_id>/objects/: JSON files defining multiple `ModelRules` for a JSON Object within a column

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -302,69 +330,251 @@ Record the version of the FOCUS specification in which this rule was introduced.
+- Open a branch with the source of git feature branch for the development work in progress for your development work.
 - Pull your branch to your development environment and perform all work specific to this AI in this branch.
+- Navigate to the appropriate model version directory under `specification/requirements_model/releases/X.X/` (or use the `latest` symlink for current development)
 - Add a file into the relevant folder `model_rules/attributes/` or `model_rules/columns/` with name `entity-name`.json (example: availabilityzone.json)
```
```suggestion
- Add a file into the relevant folder `model_rules/attributes/` or `model_rules/<dataset_id>/columns/` with name `entity-name`.json (example: availabilityzone.json)
```

mike-finopsorg replied:
```suggestion
- Add a file into the relevant folder `model_rules/attributes/` or `model_rules/datasets/<dataset_id>/columns/` with name `entity-name`.json (example: availabilityzone.json)
```

shawnalpay replied:
Yes... sorry.

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,82 @@ The model document for FOCUS contains the following major sections:
-- `Status` - Status of the rule (Valid values: Active, Depreciated)
+- `Order` - The order in which this rule should be processed or displayed
+- `DatasetType` - The dataset type this rule applies to (e.g. "CAU" for Cost and Usage, "CCT" for Contract Commitment)
+- `DatasetId` - The identifier of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "CostAndUsage" for Cost and Usage, "ContractCommitment" for Contract Commitment)
```
```suggestion
- `DatasetId` - The identifier of the dataset this rule belongs to (Required for Dataset, Column, and Object entity types, e.g. "CostAndUsage" for Cost and Usage, "ContractCommitment" for Contract Commitment)
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
````diff
@@ -302,69 +330,251 @@ Record the version of the FOCUS specification in which this rule was introduced.
+Complex rule depending on multiple prerequisites (ordered by Order field):
+```json
+"Dependencies": [
+  "CAU-BilledCost-C-000-M",        // Order: 10
+  "CAU-BillingCurrency-C-000-M",   // Order: 20
+  "CAU-ChargeCategory-C-000-M"     // Order: 30
````
I know this reads alright, but comments aren't allowed in JSON.

```suggestion
  "CAU-BilledCost-C-000-M",
  "CAU-BillingCurrency-C-000-M",
  "CAU-ChargeCategory-C-000-M"
```

mike-finopsorg replied:
The comments are there to highlight the order keys. Happy to remove but it's a little less informative without them.

shawnalpay replied:
I know... I just want to confirm that we're comfortable with carrying incorrect JSON forward for the purposes of human readability.  It highlights as red in GitHub, and the reader might start to question our collective respect of JSON syntax.  I'm okay with it if you are.

mike-finopsorg replied:
Moved them into a table.

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -145,154 +103,224 @@ This ensures traceability, uniqueness, and clarity.
+- `ColumnID`: The column identifier in UpperCamelCase
+- `EntityType`, `NNN`, `Level`: Same as previously defined
+
+The `NNN` numbering is only unique within each dataset namespace. For example, `CAU-ListPrice-C-000-M` and `CCT-ListPrice-C-000-M` can both exist independently - the `000` is reused and only needs to be unique within that specific dataset abbreviation.
```
No such column `ListPrice`.

```suggestion
The `NNN` numbering is only unique within each dataset namespace. For example, `CAU-ListUnitPrice-C-000-M` and `CCT-ListUnitPrice-C-000-M` can both exist independently - the `000` is reused and only needs to be unique within that specific dataset abbreviation.
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -302,69 +330,251 @@ Record the version of the FOCUS specification in which this rule was introduced.
-Indicate whether the rule is `active`, `deprecated`, or `reserved` for future use.
-
-**Reasoning Rules**
+Indicate whether the rule is `Active`, `Deprecated`, or `Removed` for future use.
```
```suggestion
Indicate whether the rule is `Active`, `Deprecated`, or `Removed`.
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,82 @@ The model document for FOCUS contains the following major sections:
+- `DatasetName` - The human-readable name of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "Cost and Usage" for Cost and Usage)
+- `ValidationCriteria` - The detailed criteria that defines how this rule is to be validated,
+  - `MustSatisfy` - The normative text that this rule defines
+  - `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `RECOMMENDED`, `SHOULD`, `MAY`, `OPTIONAL`)
```
We aren't using RECOMMENDED anymore, and we've never used OPTIONAL.  Also feels like we should specify MUST NOT and SHOULD NOT.

(Or did we want to retain RECOMMENDED, given that it's still in use in the 1.3 model, but won't be going forward...?)

```suggestion
  - `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `MUST NOT`, `SHOULD`, `SHOULD NOT`, `MAY`)
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,82 @@ The model document for FOCUS contains the following major sections:
-- `DatasetId` - The Id of the dataset entity this rule is related to
-- `DatasetName` - The Name of the dataset entity this rule is related to
+- `Reference` - The Column/Attribute Id this rule applies to
+- `EntityType` - The type of entity this rule applies to (Valid types: `Dataset`, `Column`, `Attribute`, `Object`, `Metadata`)
```
We're not supporting Metadata yet.  So either strike it here, or (much harder) add it to the RM.

```suggestion
- `EntityType` - The type of entity this rule applies to (Valid types: `Dataset`, `Column`, `Attribute`, `Object`)
```

mike-finopsorg committed:
Update guidelines/writing-requirements-model-guidelines.md

Co-authored-by: Shawn Alpay <shawn@finops.org>

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -495,6 +708,7 @@ Common rule for columns with a MUST be one of the allowed values requirement.
     "DatasetId": "CostAndUsage",
     "DatasetName": "Cost and Usage",
     "Type": "Static",
+    "Order": 10,
```
```suggestion
    "Order": 30,
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -540,6 +755,7 @@ Common rule for columns with a MUST be one of the allowed values requirement.
     "DatasetId": "CostAndUsage",
     "DatasetName": "Cost and Usage",
     "Type": "Static",
+    "Order": 10,
```
```suggestion
    "Order": 40,
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -570,6 +786,7 @@ Common rule for columns with a MUST be one of the allowed values requirement.
     "DatasetId": "CostAndUsage",
     "DatasetName": "Cost and Usage",
     "Type": "Static",
+    "Order": 10,
```
```suggestion
    "Order": 50,
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -600,6 +817,7 @@ Common rule for columns with a MUST be one of the allowed values requirement.
     "DatasetId": "CostAndUsage",
     "DatasetName": "Cost and Usage",
     "Type": "Static",
+    "Order": 10,
```
```suggestion
    "Order": 60,
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -630,6 +848,7 @@ Common rule for columns with a MUST be one of the allowed values requirement.
     "DatasetId": "CostAndUsage",
     "DatasetName": "Cost and Usage",
     "Type": "Static",
+    "Order": 10,
```
```suggestion
    "Order": 70,
```

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -642,3 +861,43 @@ Common rule for columns with a MUST be one of the allowed values requirement.
+    "DatasetId": "CostAndUsage",
+    "DatasetName": "Cost and Usage",
+    "Type": "Static",
+    "Order": 10,
```
```suggestion
    "Order": 80,
```

mike-finopsorg committed:
Apply suggestions from code review

Co-authored-by: Shawn Alpay <shawn@finops.org>

mike-finopsorg committed:
Make target entity a list
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Fixed up structure docs
Moved order values out of JSON to keep it valid.
Signed-off-by: Mike Fuller <mike@finops.org>

jpradocueva reviewed (approved):
Ready for Review & Approval. Thanks, @mike-finopsorg

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,89 @@ The model document for FOCUS contains the following major sections:
+- `Order` - The order in which this rule should be processed or displayed
+- `DatasetType` - The dataset type this rule applies to (e.g. "CAU" for Cost and Usage, "CCT" for Contract Commitment)
+- `DatasetId` - The identifier of the dataset this rule belongs to (Required for Dataset, Column, and Object entity types, e.g. "CostAndUsage" for Cost and Usage, "ContractCommitment" for Contract Commitment)
+- `DatasetName` - The human-readable name of the dataset this rule belongs to (Required for Column and Dataset entity types, e.g. "Cost and Usage" for Cost and Usage)
```
To match `DatasetId` immediately above.

```suggestion
- `DatasetName` - The human-readable name of the dataset this rule belongs to (Required for Dataset, Column, and Object entity types, e.g. "Cost and Usage" for Cost and Usage)
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -302,69 +337,258 @@ Record the version of the FOCUS specification in which this rule was introduced.
 
+#### 14. DatasetId and DatasetName – Dataset Association
+
+For Column and Dataset entity types, these fields establish the relationship between the rule and its parent dataset.
```
```suggestion
For Dataset, Column, and Object entity types, these fields establish the relationship between the rule and its parent dataset.
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,89 @@ The model document for FOCUS contains the following major sections:
+| Entity             | Description                                         | Applies To                                | Example RM Function                                                                                             |
 |--------------------|-----------------------------------------------------| ----------------------------------------- |-----------------------------------------------------------------------------------------------------------------|
 | `Dataset`          | Whole billing dataset                               | Structural presence, versioning, coverage | Dataset MUST include all columns required by the declared FOCUS version                                         |
 | `Row`              | Individual line item in dataset                     | Logic conditions, nullability, alignment  | Rows with `ChargeCategory = Purchase` MUST contain a `SkuId`                                                    |
```
I believe this "row" needs to go.  😏 

```suggestion
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,89 @@ The model document for FOCUS contains the following major sections:
+<img width="491" height="491" alt="Image" src="https://github.com/user-attachments/assets/a30d828e-d2af-4185-984c-475998466437"/>
 
-- **Service Provider** is not a structural entity but is frequently used as a **conditional input** to determine when a requirement applies.
+- **Dataset, Column, Attribute, Metadata** are the **core structural entities** where model requirements are directly assigned.
```
```suggestion
- **Dataset, Column, Object, and Attribute** are the **core structural entities** where model requirements are directly assigned.
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -145,154 +110,224 @@ This ensures traceability, uniqueness, and clarity.
 - Identify the first normative keyword present in the requirement:
   - `MUST`, `MUST NOT` → Mandatory
   - `SHOULD`, `SHOULD NOT` → Optional (unless conditional)
   - `MAY`, `MAY NOT` → Optional
```
```suggestion
  - `MAY` → Optional
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,89 @@ The model document for FOCUS contains the following major sections:
-
-- `ModelRuleId` - Is the formal Id given to this entry in the model rules (See: [CR Expression Format](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/1121-ai-align-on-approach-for-scrs/specification/requirements_model/README.md#-cr-expression-format))
+- `ModelRuleId` - Formal identifier for this model rule entry
 - `Function` - The type of rule to be defined (Valid types: `Composite`, `Presence`, `Type`, `Format`, `Validation`)
```
```suggestion
- `Function` - The type of rule to be defined (Valid types: `Composite`, `Presence`, `Type`, `Format`, `Validation`, `Nullability`, `Ambiguous`)
```

mike-finopsorg replied:
```suggestion
- `Function` - The type of rule to be defined (Valid types: `Composite`, `Presence`, `Type`, `Format`, `Validation`, `Nullability`)
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,126 +18,89 @@ The model document for FOCUS contains the following major sections:
+  - `MustSatisfy` - The normative text that this rule defines
+  - `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `MUST NOT`, `SHOULD`, `SHOULD NOT`, `MAY`)
+  - `Requirement` - The definition of what is required for model
+  - `Condition` - The definition of conditions under which this rule applies
```
```suggestion
  - `Condition` - The definition of conditions under which this rule applies
  - `Dependencies` - A list of prerequisite rules that must be evaluated before this rule, ordered by sequence
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -145,154 +110,224 @@ This ensures traceability, uniqueness, and clarity.
-
-- Use the `display_name` for the column as written in the spec.
-- For rules related to attribute-level constraints (e.g., `NumericFormat`), use the attribute name.
-- This field should exactly match the title of the column or attribute from the normative requirements.
+- For columns: Use the ColumnId (e.g., `CommitmentDiscountQuantity`)
+- For attributes: Use the attribute name (e.g., `NumericFormat`, `StringHandling`)
```
How should this get populated for Dataset and Object?  I would provide a suggestion, but I'm not actually sure...

mike-finopsorg replied:
```suggestion
- For datasets: Use the DatasetId (e.g., `CostAndUsage`)
- For columns: Use the ColumnId (e.g., `CommitmentDiscountQuantity`)
- For objects: Use the ObjectId (e.g., `AllocatedMethodDetailsObject`)
- For attributes: Use the attribute name (e.g., `NumericFormat`, `StringHandling`)
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -145,154 +110,224 @@ This ensures traceability, uniqueness, and clarity.
-  `Requirement = NumericFormat:CR`
-- Set to `"null"` for atomic rules that do not depend on other rules or attributes.
-- Always define this field for composite rules, and make sure referenced CRIDs exist or appear later in the table.
+For **Composite rules**, use logical CheckFunctions to group other rules:
+- `CheckFunction: "AND"` with `Items` array containing nested requirement references
+- `CheckFunction: "OR"` with `Items` array containing nested requirement references
```
```suggestion
For **Composite rules**, use logical CheckFunctions to group other rules:
- `CheckFunction: "AND"` with `Items` array containing nested requirement references
- `CheckFunction: "OR"` with `Items` array containing nested requirement references
- Use `CheckFunction: "CheckModelRule"` inside the `Items` array to reference other RMIDs
```

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -145,154 +110,224 @@ This ensures traceability, uniqueness, and clarity.
+For **Atomic rules**, use specific CheckFunction types from the model:
+- `ColumnPresent` - Check if a column exists in the dataset
+- `CheckNotValue` - Verify a column does not contain a specific value
+- `CheckValue` - Verify a column contains a specific value
```
```suggestion
- `CheckValue` - Verify a column contains a specific value
- `CheckAllowedValues` - Verify a column contains one of the allowed values
```

mike-finopsorg committed:
Use X.Y instead of X.X for versions
Remove entity Row this is incorrect
Signed-off-by: Mike Fuller <mike@finops.org>

beau-nelford-anglepoint reviewed (approved):
Approving. All of Shawn's March 1 feedback has been addressed in Mike's March 4 updates. The Requirements Model guidelines are now clear and comprehensive with proper RM terminology, multi-dataset namespacing, and updated examples throughout.

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -137,154 +110,224 @@ This ensures traceability, uniqueness, and clarity.
+- Use `Nullability` to define conditional nullability rules (e.g., "MUST NOT be null when condition X" or "MAY be null when condition Y").
+- Use `Validation` for business logic or fixed-value conditions not covered above, including unconditional NOT NULL rules.
+- Use `Composite` to group multiple RMIDs with logical expressions (`AND` / `OR` / `NOT`).
 - Use `Ambiguous` only when no clear classification is possible.
```
```suggestion
```

mike-finopsorg committed:
Apply suggestions from code review

Co-authored-by: Shawn Alpay <shawn@finops.org>

mike-finopsorg commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -16,118 +18,89 @@ The model document for FOCUS contains the following major sections:
+| Entity             | Description                                         | Applies To                                | Example RM Function                                                                                             |
 |--------------------|-----------------------------------------------------| ----------------------------------------- |-----------------------------------------------------------------------------------------------------------------|
 | `Dataset`          | Whole billing dataset                               | Structural presence, versioning, coverage | Dataset MUST include all columns required by the declared FOCUS version                                         |
 | `Row`              | Individual line item in dataset                     | Logic conditions, nullability, alignment  | Rows with `ChargeCategory = Purchase` MUST contain a `SkuId`                                                    |
```
```suggestion
```

thecloudman commented on specification/requirements_model/model_rules/attributes/columnhandling.json:
"Custom columns MUST be prefixed with a consistent x_ prefix to identify them as external  or custom columns and distinguish them from FOCUS columns to avoid conflicts in future releases.",

thecloudman replied:
removed , after external change to or maybe

shawnalpay replied:
The way it's quoted here is the way it's written in the [attribute](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/attributes/column_handling.md).  Garbage in garbage out... 🚮 

> Custom columns MUST be prefixed with a consistent x_ prefix to identify them as external, custom columns and distinguish them from FOCUS columns to avoid conflicts in future releases.

mike-finopsorg commented:
Hey All,

This PR has a lot of content pulled into it as it was originally forked from the 1.3 development branch which is now merged into working_draft. To avoid the large amount of conflicts which are not relevant to the proposed change here I have opened a new PR which just has the guideline updates here: https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2051/files

Before I close this PR out I wanted to ensure I have resolved the previously open comments people had on this PR which I have resolved here and taken over to the new PR.

shawnalpay commented:
> Before I close this PR out I wanted to ensure I have resolved the previously open comments people had on this PR which I have resolved here and taken over to the new PR.

@mike-finopsorg I believe we are good to close.  Setting Status to `Closed` so the WG knows not to review this PR.
